wagtail/wagtailadmin/edit_handlers.py

```diff
@@ -95,59 +96,85 @@ class EditHandler(object):
     the EditHandler API
     """
 
+    def __init__(self, heading='', classname='', help_text=''):
+        self.heading = heading
+        self.classname = classname
+        self.help_text = help_text
+
+    def clone(self):
+        return self.__class__(
+            heading=self.heading,
+            classname=self.classname,
+            help_text=self.help_text,
+        )
+
     # return list of widget overrides that this EditHandler wants to be in place
     # on the form it receives
-    @classmethod
-    def widget_overrides(cls):
+    def widget_overrides(self):
         return {}
 
     # return list of fields that this EditHandler expects to find on the form
-    @classmethod
-    def required_fields(cls):
+    def required_fields(self):
         return []
 
     # return a dict of formsets that this EditHandler requires to be present
     # as children of the ClusterForm; the dict is a mapping from relation name
     # to parameters to be passed as part of get_form_for_model's 'formsets' kwarg
-    @classmethod
-    def required_formsets(cls):
+    def required_formsets(self):
         return {}
 
     # return any HTML that needs to be output on the edit page once per edit handler definition.
     # Typically this will be used to define snippets of HTML within <script type="text/x-template"></script> blocks
     # for Javascript code to work with.
-    @classmethod
-    def html_declarations(cls):
+    def html_declarations(self):
         return ''
 
-    def __init__(self, instance=None, form=None):
+    def bind_to_model(self, model):
+        new = self.clone()
+        new.model = model
+        new.on_model_bound()
+        return new
+
+    def on_model_bound(self):
+        pass
+
+    def bind_to_instance(self, instance=None, form=None):
+        new = (self.bind_to_model(self.model) if hasattr(self, 'model')
+               else self.clone())
```

Comment:
I'm a bit lost about what's happening here. Is `self.bind_to_model(self.model)` really just saying "create a clone of this EditHandler that preserves its `model` property"? If so, let's create a new method to make that explicit. (In fact, can't we just make it the `clone` method's responsibility to copy the `model` attribute whenever it exists?)

 Also, is the `if hasattr(self, 'model')` test really needed here? Is there any circumstance where you would call `bind_to_instance` on an edit handler that _hasn't_ been bound to a model at that point?

Reply:
Thanks for the review! I know this is pretty complex, due to the laziness we need in the model+instance binding… For these two particularly reasons, it was designed like this on purpose.

There are two reasons why I don’t assign `model` in `.clone()`:
- Simplicity: it would need to add `new = self.__class__(…)` + `new.model = self.model` + `return new` in all `clone` methods. Since this method is often changed in `EditHandler` subclasses (including third-party ones), it’s better like this. Of course, we can’t add `model` as an `EditHandler.__init__(…)` kwarg, or users would have a weird `model` argument available in `FieldPanel(…)` etc.
- The `on_model_bound` method. This method is really mandatory and can’t be put in `__init__` since the model is always bound to edit handlers **after** their initialization. If we want to bind models in `.clone()`, we have to do this:
```python
def __init__(self, …):
    …
    self.model = None

def clone(self):
    new = self.__class__(…)
    new.model = self.model
    if self.model is not None:
        new.on_model_bound()
    return new

def bind_to_model(self, model):
    new = self.clone()
    new.model = model
    new.on_model_bound()
    return new
```

As you can see, this makes code more redundant, and it does more copies. Also, this doesn’t fully work. I tried it, we have to adjust code in a lot of places in order to get it working (in 30 minutes of adjustments, I couldn’t make the tests pass, the model binding is partially broken).

About the `if hasattr(self, 'model')` test, you’re right, models should always be bound at this point. There is a small possibility though: when you define the `edit_handler` class attribute of your model with some panels, then modify in place the list of its children in that particular case, the model will not be bound.
But in that case, the code I wrote will not work, it’s not `self.clone()` that should be called, but `self.bind_to_model(instance._meta.model)`. This this last statement should work in all case, I think I’ll add a test for this edge case and replace that line with `new = self.bind_to_model(instance._meta.model)`.

Reply:
In fact, after testing and reading in detail the view logic, it’s not possible to bind an instance without the model bound. The reason is simple: to bind an instance we need a `form` argument. To get the form, we call `EditHandler.get_form_class(…)` that relies on the bound model.